Merging beta bot into dev

BrandonBeaver117 opened the pull request.

BrandonBeaver117 committed:
Starting for arm

Basically, porting over some of the ideas from last year's arm

BrandonBeaver117 committed:
Adding commands

commands and robot container

BrandonBeaver117 committed:
Merge branch 'dev' into feat/arm

BrandonBeaver117 committed:
Arm and Encoder Ports

Adding the real ports

BrandonBeaver117 committed:
Getting angle from joystick

Using joystick to get actual angles

BrandonBeaver117 committed:
Adding shuffleboard

Now, current and desired angle are seen

Naruto-c committed:
State machine(?)

Poorly made state machine, needs optimization

Naruto-c committed:
Ignore past commit 

Forgot to save and all it did was add a new file

Naruto-c committed:
Merge remote-tracking branch 'origin/feat/outtake' into feat/outtake

Naruto-c committed:
Added enums and state machine

State machine is very very badly made, will probably need optimization

Naruto-c committed:
Updated some comments

Naruto-c committed:
Added the extra motor ports and values to work with

two more motors :`(

Naruto-c committed:
Added constants and renamed intake and outtake

Intake --> Grab
Outtake --> Release

Naruto-c committed:
Working periodics

Hopefully, the periodics will work after passing in actual values in the constants file

BrandonBeaver117 committed:
Really only needing two states

Open and clamp!

BrandonBeaver117 committed:
Don't need to lock mode really

BrandonBeaver117 committed:
Outtake for real

Uses motor angle and high pass filters to move claw outtake mechanism. Not tested yet

BrandonBeaver117 committed:
End of day

It works as intended, though might be some engineering things to fix up to actually make it grab

BrandonBeaver117 committed:
Using Low pass filter

BrandonBeaver117 committed:
Adding good stator limits

audreypj committed:
Adding telescoping arm subsystem into arm subsystem

audreypj committed:
adding a couple extension safety methods

audreypj committed:
Adding telescoping arm command with safety

BrandonBeaver117 committed:
More tunings for arm 

Adds in some tunings after testing. Also, added a check if we are in a "finished"  state before switching states

BrandonBeaver117 committed:
Spacing

BrandonBeaver117 committed:
formatting??

audreypj committed:
Safety methods

audreypj committed:
Merge branch 'dev' into integrate/arm-telescoping

audreypj committed:
Creating and binding arm position command

BrandonBeaver117 committed:
Merge branch 'dev' into feat/outtake

audreypj committed:
minor style improvements

BrandonBeaver117 committed:
Bit more tunings and cleaning

BrandonBeaver117 committed:
Better tunings after testing

woo, it kinda works

audreypj committed:
debugging/testing

BrandonBeaver117 committed:
Commands turn off when let go?

BrandonBeaver117 committed:
Commands switch when let go?

BrandonBeaver117 committed:
Merge remote-tracking branch 'origin/feat/outtake' into feat/outtake

BrandonBeaver117 committed:
Better Modes

Stator currents now defined within constants, and creating modes with those currents in the outtake subsystem. Better naming and optionals also used

BrandonBeaver117 committed:
While true

Allows the mode to be set, but when the button is let go, the ending mode can b set

BrandonBeaver117 committed:
Better commands

Sets it to the proper state if we fail

JayK445 committed:
gravity control fixes?

set the gravity offset from cos to sin, increased clamp range, made gravity constant larger and positive

BrandonBeaver117 committed:
Simpler commands

Write faster, less complexity.

JayK445 committed:
remove duplicated code

remove duped logic, created method

BrandonBeaver117 committed:
Nicer function

BrandonBeaver117 committed:
Adding back the telescoping arm controls

BrandonBeaver117 committed:
Clamping

BrandonBeaver117 committed:
Better manual command

BrandonBeaver117 committed:
Deadbanding

BrandonBeaver117 committed:
Tuning PID

Arm works pretty well, needs more tuning to dampen osillcaion

BrandonBeaver117 committed:
Tuning extension pid

Glass why u do this

Craptop committed:
make the logic slightly better

a-viv-a committed:
make function names more descriptive

a-viv-a committed:
refactor arm subsystem


rename function again


rearrange and refactor arm code further

Craptop committed:
fix compile errors with scoping and reduce p term

Craptop committed:
stop using the wrong constant for clamping

Craptop committed:
dont need to add 0+1

Craptop committed:
inverting arm encoder

Craptop committed:
reverting to previous GradleRIO version

Craptop committed:
tuning max extension value and extension PID

Craptop committed:
Merge branch 'integrate/arm-telescoping' into integration/beta-bot

Craptop committed:
Merge branch 'feat/outtake' into integration/beta-bot

Craptop committed:
updating button bindings

Craptop committed:
set intaking button to max extension

audreypj committed:
Add button bindings, manual extension, position command ends, constants

audreypj committed:
update atTarget method to compare current and target values

a-viv-a committed:
fix spotless formatting issues

audreypj committed:
fix layer button bindings, improve constants, combine manual commands

BrandonBeaver117 committed:
Merge remote-tracking branch 'origin/integration/beta-bot' into feat/outtake

Craptop committed:
Merge branch 'integration/beta-bot' into integrate/arm-telescoping

Craptop committed:
Merge branch 'integration/beta-bot' into integrate/arm-telescoping

Craptop committed:
tuning, addrequirements to outtake commands

Craptop committed:
adding return arm to original position button binding

Craptop committed:
remove unneeded line

BrandonBeaver117 committed:
Timed outtake

Craptop committed:
zeroing

Craptop committed:
make atTarget work so manual command can run

BrandonBeaver117 committed:
Merge branch 'integration/beta-bot' into feat/outtake

Craptop committed:
Merge branch 'integration/beta-bot' into integrate/arm-telescoping

BrandonBeaver117 committed:
Spotless happy

BrandonBeaver117 committed:
Force intake

Not sure how clean the mode finished method is, but tried to match AlphaBot code. Maybe we also wanna do force outtake for future?

BrandonBeaver117 committed:
Locking mode

BrandonBeaver117 committed:
Reliable outtake

Getting rid of the BS

BrandonBeaver117 committed:
Merge branch 'feat/outtake' into integration/beta-bot

Craptop committed:
Zeroing (hopefully)

Craptop committed:
changed how drivetopos mode gets set

Craptop committed:
changes how setzeromodecommand ends

Craptop committed:
apply gravity constant

a-viv-a committed:
Merge remote-tracking branch 'origin/dev' into integration/beta-bot

Craptop committed:
debugging zero mode

Craptop committed:
capping sequencer

Craptop committed:
change arm ports

Craptop committed:
changed outtake motor port

Craptop committed:
new offset and max extension

Craptop committed:
Merge branch 'integrate/arm-telescoping' into integration/beta-bot

Craptop committed:
Tuning arm

Bigger P-term while also switching to profiled PID controller

Craptop committed:
tune arm pid

BrandonBeaver117 committed:
Merge remote-tracking branch 'origin/dev' into integration/beta-bot

BrandonBeaver117 committed:
Getting rid of unnecessary modes import

BrandonBeaver117 commented:
Some features (like the arm) might be tuned with c-bot, not just beta-bot

BrandonBeaver117 committed:
C-Bot arm extension testing

BrandonBeaver117 committed:
Getting rid of unneeded comments in constants

Some values persisted that are unneeded

BrandonBeaver117 committed:
Death to the terrible outtake end command

Logic doesn't make sense, and it's very unnecessary. Might bring it back in modes in the outtake subsystem...but better!

audreypj committed:
Delete TelescopingArmPositionCommand.java

not necessary

audreypj reviewed (approved):
Looks good to me, I deleted an unnecessary telescoping arm command, Brandon made some small edits to constants and deleted another unnecessary command. Lexie reviewed with me.

Craptop committed:
fix the number of leds

Craptop committed:
Merge remote-tracking branch 'origin/integration/beta-bot' into integration/beta-bot

Craptop committed:
formatting and test red command

Craptop committed:
nicer yellow

Craptop committed:
show colors based on network health

Craptop committed:
new mint color

Craptop committed:
unflip motor

a-viv-a commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -135,6 +137,87 @@ public static final class BackRight { // Module 4
+    public static final double ANGULAR_OFFSET = -8.75;
+
+    public static final class Setpoints {
+      public static final class ScoreLow {
```
I think this will need refactoring into an object, so that it can be bundled into an object that encompasses a pose and arm state for scoring, but this refactor can come later.

a-viv-a commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -135,6 +137,87 @@ public static final class BackRight { // Module 4
+
+    public static final double EXTENSION_STATORLIMIT = 80;
+
+    public static final double ZERO_RETRACTION_PERCENT = -0.4; // FIXME
```
This value is possibly too high.

a-viv-a commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -190,6 +273,37 @@ public static final class PoseEstimator {
+      public static final int OUTTAKE_MOTOR = 8; // Placeholder value
+    }
+
+    public static final int OPEN_ANGLE = 500;
```
What is this value used for? (nothing)

a-viv-a commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -190,6 +273,37 @@ public static final class PoseEstimator {
+    }
+
+    public static final int OPEN_ANGLE = 500;
+    public static final int CLAMP_ANGLE = 0;
```
same as above

a-viv-a commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -190,6 +273,37 @@ public static final class PoseEstimator {
+
+    // Thinking of using these to plug into the stator limits above...?
+    // Better readability?
+    private static final class StatorCurrents {
```
are these used? they are marked private which is wrong if they are used

a-viv-a commented on src/main/java/frc/robot/commands/OuttakeCommand.java:
```diff
@@ -0,0 +1,41 @@
+import frc.robot.subsystems.OuttakeSubsystem;
+import frc.robot.subsystems.OuttakeSubsystem.Modes;
+
+public class OuttakeCommand extends CommandBase {
```
this command should be renamed to SetOuttakeSubsystemModeCommand for clarity--its long, but it will disambiguate with the intake/system state better.

a-viv-a commented on src/main/java/frc/robot/commands/ForceOuttakeCommand.java:
```diff
@@ -0,0 +1,42 @@
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.subsystems.OuttakeSubsystem;
+
+public class ForceOuttakeCommand extends CommandBase {
```
This command should be renamed to ForceOuttakeSubsystemModeCommand for same reasons as below. Use f2 for mass rename!

a-viv-a commented on src/main/java/frc/robot/subsystems/OuttakeSubsystem.java:
```diff
@@ -0,0 +1,164 @@
+      this.outtakeDetails = outtakeDetails;
+    }
+
+    public boolean modeFinished(
```
why is this public? it doesnt make sense as something a command should be allowed to control

BrandonBeaver117 replied:
Might just be my poor programming knowledge, but if the method was private, couldn't the method only be called within the class? Don't we want to expose this method to the advance periodic to see if the mode is finished?

a-viv-a replied:
This is certainly an edgecase, you don't need to feel bad about not knowing it. Your understanding of private is correct! Making the nested class method private means it can only be accessed within the class... which is your outtakesubsystem.

https://docs.oracle.com/javase/specs/jls/se7/html/jls-6.html#jls-6.6.1:~:text=Otherwise%2C%20if%20the%20member%20or%20constructor%20is%20declared%20private%2C%20then%20access%20is%20permitted%20if%20and%20only%20if%20it%20occurs%20within%20the%20body%20of%20the%20top%20level%20class%20(%C2%A77.6)%20that%20encloses%20the%20declaration%20of%20the%20member%20or%20constructor.

Private is scoping the method to only be callable within the top level class. Nested classes are supposed to be an organizational tool, not an access boundary. If you need to hide implementation details of your nested class from your higher level class, you need to create a new class. For this reason, classes within a top level class share access restrictions with their top level class, and the inverse. Nested class can access all privates, and upper class can do the same--they are essentially one class with some namespacing.

If you ever want to play with access control, try changing the access of a method or member and see what happens! In this case, it works fine, which you can see if you try building after the change.

a-viv-a reviewed (changes requested):
Largely looking great, let's make the cleanup changes and merge this today.

BrandonBeaver117 committed:
Better names

Commands have more descriptive and apt names. Also changed scoping within outtake subsystem to be better

BrandonBeaver117 committed:
Spotless fixes

a-viv-a reviewed (approved):
this isnt perfect, but its pretty good--lets merge it and make it better later (copium)
